quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -0,0 +1,824 @@
+/// Complex machinery for simplifying [`Expression`]s.
+use crate::expression::{
+    imag, real, Expression, ExpressionFunction, FunctionCallExpression, InfixExpression,
+    InfixOperator, PrefixExpression, PrefixOperator,
+};
+use std::cmp::min_by_key;
+
+/// Simplify an [`Expression`].
+pub(super) fn run(expression: &Expression) -> Expression {
+    simplify(expression, LIMIT)
+}
+
+/// Keep stack sizes under control
+const LIMIT: u64 = 10;
```

Comment:
How did you choose 10?  What would happen if the limit were greater by an order of magnitude?

Reply:
I played around with the number while looking at various test cases. One that's commented out, the _big_ one, will grow unbounded and crash with an "I've overflowed my stack" message if `LIMIT` is set to 100. I think the simpler cases might end up doing that around 100 as well? If they can't simplify further, we just build up a huge stack trace through multiple recursive calls and then it overflows.

Reply:
I'll leave a comment.

Reply:
FYI: https://doc.rust-lang.org/reference/attributes/limits.html (maybe 10 is fine though...)

Reply:
Looks like that's for compile-time operations, though?

Reply:
Oops, you're right.